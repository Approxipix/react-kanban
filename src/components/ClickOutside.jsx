import React, { Component } from "react";

class ClickOutside extends Component {
  componentDidMount() {
    document.addEventListener('click', this.outerClick);
  }

  componentWillUnmount() {
    document.removeEventListener('click', this.outerClick);
  }

  outerClick = (e) => {
    let { target } = e;
    if (!!target.closest && (target.closest('.click-outside-wrapper'))) {
      return;
    }
    this.props.toggleOpened();
  };

  render() {
    return (
      <div className="click-outside-wrapper">
        {this.props.children}
      </div>
    )
  };
}

export default ClickOutside;
