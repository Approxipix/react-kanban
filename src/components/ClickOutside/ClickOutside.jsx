import React, { Component } from "react";
import PropTypes from 'prop-types';
import styled from 'styled-components';

const Wrapper = styled.div``;
Wrapper.displayName = 'Wrapper';

class ClickOutside extends Component {
  componentDidMount() {
    const { action } = this.props;
    document.addEventListener(!!action ? action : 'click', this.outerClick);
  }

  componentWillUnmount() {
    const { action } = this.props;
    document.removeEventListener(!!action ? action : 'click', this.outerClick);
  }

  outerClick = (e) => {
    let { target } = e;
    if (!!target.closest && (target.closest('.click-outside-wrapper'))) {
      return null;
    }
    this.props.toggleOpened();
  };

  render() {
    return (
      <Wrapper className="click-outside-wrapper">
        {this.props.children}
      </Wrapper>
    )
  };
}

ClickOutside.defaultProps = {
  toggleOpened: () => {}
};

ClickOutside.propTypes = {
  toggleOpened: PropTypes.func.isRequired
};

export default ClickOutside;
