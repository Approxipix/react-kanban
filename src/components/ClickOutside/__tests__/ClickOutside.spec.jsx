import React from 'react';
import { shallow } from 'enzyme';
import toJson from "enzyme-to-json";
import ClickOutside from '../ClickOutside';

describe('<ClickOutside>', () => {
  let component;
  let mockToggleOpened;

  beforeEach(() => {
    mockToggleOpened = jest.fn();
    const props = {
      toggleOpened: mockToggleOpened
    };
    component = shallow(
      <ClickOutside {...props}>
        <p>Test Component</p>
      </ClickOutside>
    );
  });

  it('should renders correctly', () => {
    expect(toJson(component)).toMatchSnapshot()
  });

  it('should not call toggleOpened on click inside the component', () => {
    component.find('Wrapper').simulate('click');

    expect(mockToggleOpened).toHaveBeenCalledTimes(0)
  })
});
