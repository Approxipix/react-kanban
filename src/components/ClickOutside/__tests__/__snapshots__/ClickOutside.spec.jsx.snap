// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ClickOutside> should renders correctly 1`] = `
<Wrapper
  className="click-outside-wrapper"
>
  <p>
    Test Component
  </p>
</Wrapper>
`;
